helix-rest/src/main/java/org/apache/helix/rest/common/datamodel/RestSnapShot.java

```diff
@@ -19,18 +18,69 @@
  * under the License.
  */
 
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.helix.HelixProperty;
+import org.apache.helix.PropertyKey;
+import org.apache.helix.PropertyType;
+
 /* This Snapshot can extend Snapshot from common/core module
  * once there is more generic snapshot.
+ * An Snapshot object should contain all the Helix related info that an implementation of
+ * OperationAbstractClass would need.
  */
+
+// TODO: Future: Support hierarchical Snapshot type for other services besides cluster MaintenanceService.
+
 public class RestSnapShot {
-  /* An Snapshot object should contain all the Helix related info that an implementation of
-   * OperationAbstractClass would need.
-   */
+  private final Map<PropertyKey, HelixProperty> _propertyCache;
+  private final Map<PropertyKey, List<String>> _childNodesCache;
+  private Set<PropertyType> _propertyTypes;
+  private String _clusterName;
+
+  public RestSnapShot(String clusterName) {
+    _propertyCache = new HashMap<>();
+    _childNodesCache = new HashMap<>();
+    _propertyTypes = new HashSet<>();
+    _clusterName = clusterName;
+  }
+
+  public <T extends HelixProperty> T getProperty(PropertyKey key) {
```

Comment:
Do we support this kind of return type in SNAPSHOT for RoutingTable? If yes, let's keep this convention. Otherwise, this is more specific for snapshot public API.